### PR TITLE
raft: remove Node interface and impl

### LINKED ISSUE #122131: raft: repurpose unused Node interface
We only use [RawNode](https://github.com/cockroachdb/cockroach/blob/e37b1f17ce531822e96d095bd32db0acde903e73/pkg/raft/rawnode.go#L37) interface to `raft`, and don't use the [Node](https://github.com/cockroachdb/cockroach/blob/e37b1f17ce531822e96d095bd32db0acde903e73/pkg/raft/node.go#L129) which is a wrapper around `RawNode` with goroutines and channels that runs the node. In CRDB, we do all that in the raft scheduler.

We could outright remove the `Node` interface, but maybe we should leave if for additional testing coverage, or migrate the tests to use `RawNode`. The `Node` tests tend to be flaky (e.g. #121745) because of timing/scheduling: e.g. a test may wait for a leader and then assume that the leader is stable, but timing flakes can result in a leader change in the middle of the test.

A better approach to testing would be to have randomized, but deterministic tests based on `RawNode`. Then all the flakes would be preventable and/or reproducible.

We probably can repurpose `Node` to be a test-only type, with added benefits of randomized testing. We could improve reproducibility of these runs, by tracing all `Step/Ready` calls and message sends/drops done by the `Node`s. The trace would be convertible to a list of `datadriven` commands that `TestInteraction` operates with, so any test run could be then reliably reproduced with `RawNode`s.

Jira issue: CRDB-37718

Epic CRDB-39898

### PR DESCRIPTION
**Pull request by pav-kv**
The `Node` interface and implementation was used only in tests. Most of these tests were deterministic, so can be done with a `RawNode`. A few tests in `rafttest` are more of a "chaos" tests - for them, there is a minimal `rafttest.node` implementation which serves its purposes.

After this PR, there is no longer `raft.Node` API to maintain.

Resolves #122131

### CODE REVIEW

**Comment by pav-kv** [comment]
@tbg @nvanbenschoten PTAL

The first 18 commits migrate tests in `raft` package to `RawNode`, and are fairly mechanical. You might want to review each or a few of those commits to get the idea, or look at them wholesale in Reviewable, or skip if you believe me the transition is a no-op.

The 19th commit refactors the harness in `raft/rafttest` to use `RawNode` instead of `Node`, and retains some non-determinism (so these are more like "chaos" tests). It's the only non-deterministic harness we now have in `raft` package (I think). It mimics the CRDB's use: a `Ready` loop running in parallel with `Step`s.

The last commit is my favourite.

There are a few lint checks failing, but these should not block the review, I'll fix after your review before merging.

File: pkg/raft/node.go [hunk 1]
```diff
@@ -124,468 +114,15 @@ func IsEmptySnap(sp pb.Snapshot) bool {
 	return sp.Metadata.Index == 0
 }
 
-// Node represents a node in a raft cluster.
-type Node interface {
-	// Tick increments the internal logical clock for the Node by a single tick. Election
-	// timeouts and heartbeat timeouts are in units of ticks.
-	Tick()
```
Comment (pav-kv) [inline pkg/raft/node.go:131]: Might want to double-check that the `RawNode` comments for these methods are not missing details that might be present in this interface's comments, and "migrate" the missing bits over if any.

**Review by tbg — APPROVED** [review]
❤️ 
meme

File: pkg/raft/node.go (hunk 1 above)
Comment (pav-kv) [reply pkg/raft/node.go:131]: There are indeed a bunch of comments that can be copied over to `RawNode`. I'll send another PR with updated `RawNode` comments, will need to revisit each one.

**Comment by pav-kv** [comment]
bors r=tbg

File: pkg/raft/node.go (hunk 1 above)
Comment (pav-kv) [reply pkg/raft/node.go:131]: #132006

**Pull request merged**